parser/charset/encoding_table.go

```diff
@@ -298,3 +300,174 @@ func characterLengthUTF8(bs []byte) int {
 	}
 	return 4
 }
+
+func characterLengthOne(_ []byte) int {
+	return 1
+}
+
+type TruncateStrategy int8
+
+const (
+	TruncateStrategyEmpty TruncateStrategy = iota
+	TruncateStrategyTrim
+	TruncateStrategyReplace
+)
+
+var _ StringValidator = StringValidatorASCII{}
+var _ StringValidator = StringValidatorUTF8{}
+var _ StringValidator = StringValidatorOther{}
+
+// StringValidator is used to check if a string is valid in the specific charset.
+type StringValidator interface {
+	Validate(str string) (invalidPos int)
+	Truncate(str string, strategy TruncateStrategy) (result string, invalidPos int)
+}
+
+// StringValidatorASCII checks whether a string is valid ASCII string.
+type StringValidatorASCII struct{}
+
+// Validate checks whether the string is valid in the given charset.
+func (s StringValidatorASCII) Validate(str string) int {
+	_, invalidPos := s.Truncate(str, TruncateStrategyEmpty)
+	return invalidPos
+}
+
+// Truncate implement the interface StringValidator.
+func (s StringValidatorASCII) Truncate(str string, strategy TruncateStrategy) (string, int) {
+	var result []byte
+	if strategy == TruncateStrategyReplace {
+		result = make([]byte, 0, len(str))
+	}
+	invalidPos := -1
+	for i, w := 0, 0; i < len(str); i += w {
+		w = 1
+		if str[i] > go_unicode.MaxASCII {
+			if invalidPos == -1 {
+				invalidPos = i
+			}
+			switch strategy {
+			case TruncateStrategyEmpty:
+				return "", invalidPos
+			case TruncateStrategyTrim:
+				return str[:i], invalidPos
+			case TruncateStrategyReplace:
+				w = characterLengthUTF8(Slice(str)[i:])
+				w = mathutil.Min(w, len(str)-i)
+				result = append(result, '?')
+				continue
+			}
+		}
+		if strategy == TruncateStrategyReplace {
```

Comment:
Could we omit `result = make([]byte, 0, len(str))` if there are  no invalid strings.

For example, we can add a variable that only needs to be copied if `str[I] > go_unicode.MaxASCII` is present